[Modal] Apply aria-hidden to the correct Elements in the tree when Modals are mounted to to other places than document.body

Gr3q opened the pull request:
Fixes https://github.com/mui/material-ui/issues/19450

Before if someone used `disablePortal` and opened a modal the whole page became inaccessible to Screen Readers because `aria-hidden` only got applied to siblings of document.body, in this case all of them. This PR fixes that.

Changes:
* Make sure we pass in the correct container if `disablePortal` is used.
* If the modal container is deep in the tree we need to make sure every sibling in every ancestor level is aria-hidden and and every aria-hidden we added gets cleaned up properly on unmount.
*  In addition because we don't differentiate between aria-hidden tags that got added by modals and added by devs, we need to pierce through (remove) every aria-hidden attribute in every ancestor of the modal container; this is to prevent no accessible elements on a page when multiple modals are present.

I'll port this to next when I have time.

- [ ] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Gr3q commented:
You argos check seems to be borked especially I made no changed that would affect visuals.

ZeeshanTamboli reviewed (commented):
@Gr3q, can you point the PR to the `master` branch, our current development branch for releases? I'll review it once that's done.

Gr3q commented:
Should I not touch the unstable folder in the future? Could you explain why does it exists?

ZeeshanTamboli commented:
> Should I not touch the unstable folder in the future? Could you explain why does it exist?

I reverted the changes in Base UI (`mui-base` directory). It included the unstable folder because Base UI was still in beta. We no longer maintain Base UI in this repository—it's now legacy and will be removed in the future (timeline unknown). Material UI used to depend on Base UI, but we've since separated the code, causing duplication. We don't plan to make any further changes to Base UI. The issue is related to Material UI's Modal. Base UI with headless components is now being developed in a separate repo: [mui/base-ui](https://github.com/mui/base-ui), which will be integrated into Material UI once stable.

ZeeshanTamboli commented on packages/mui-material/src/Modal/ModalManager.test.ts:
```diff
@@ -425,11 +451,65 @@ describe('ModalManager', () => {
+      expect(mainContentSibling.children[0]).not.toBeAriaHidden();
+    });
+
+    it('top modal should always be accessible even if even inside other dialog', () => {
```
```suggestion
    it('top modal should always be accessible even if inside other dialog', () => {
```

ZeeshanTamboli commented on packages/mui-material/src/Modal/useModal.ts:
```diff
@@ -83,7 +84,13 @@ function useModal(parameters: UseModalParameters): UseModalReturnValue {
+     * but this will also work because the logic matches {@link Portal}.
+     */
+    const resolvedContainer = disablePortal
+      ? ((mountNodeRef.current ?? modalRef.current)?.parentElement ?? getDoc().body)
```
When `disablePortal` is true, `mountNodeRef` and `modalRef` are the same. So why should `mountNodeRef` have priority?

Gr3q replied:
naming-wise `mountNodeRef` is the node `modalRef` is mounted to, so it makes sense even if they are the same.

Plus after reading the `Portal` implementation I can see that `mountNodeRef` can still come back as `null` even if the typing for it is not indicating it.

ZeeshanTamboli commented on packages/mui-material/src/Modal/useModal.ts:
```diff
@@ -113,7 +120,11 @@ function useModal(parameters: UseModalParameters): UseModalReturnValue {
+  // aria-hidden tags have time to get cleaned up properly
+  // in handleClose->manager.remove->ariaHiddenElements
+  // in the case someone unmounts the Modal higher up the tree
+  React.useLayoutEffect(() => {
```
Are you sure layout effect is absolutely necessary here?

Gr3q replied:
Added a test for it in Modal.test.js.

ZeeshanTamboli replied:
I used our util `useEnhancedEffect` instead which handles on SSR as well.

ZeeshanTamboli commented on packages/mui-material/src/Modal/ModalManager.ts:
```diff
@@ -56,22 +56,56 @@ function isAriaHiddenForbiddenOnElement(element: Element): boolean {
+  const html = ownerDocument(container).body.parentElement;
+  const blacklist = [mountElement, ...elementsToExclude];
+
+  // In theory this should not happen anymore.
+  // in some cases the container and previous element still
+  // could end up being the same, in this case we just go up 1
+  if (current === previousElement) {
+    current = current.parentElement;
+  }
```
When could this edge case happen? Can you give a real-world example?

Gr3q replied:
It's been like 2 years since I wrote this, it can happen if the container and the current element passed in are the same, but I don't remember where it was called like that.

I think I fixed it but I kept the check in just in case so this function always exits properly no matter how it's called (it's a while loop after all).

ZeeshanTamboli replied:
That’s defensive. If the container and the starting element are the same, it means the container resolution logic is flawed. When we trust that the DOM structure is correct, so we shouldn’t include this. I am removing it.

Gr3q replied:
It was misused in the library itself before so yes, it is defensive because I can't guarantee that someone will not make the same mistake again in the future.

It's too late to modify the runtime behavior to throw/warn to defend against when someone does this - I don't want to break apps depending on material-ui after all.

ZeeshanTamboli commented on packages/mui-material/src/Modal/ModalManager.ts:
```diff
@@ -56,22 +56,56 @@ function isAriaHiddenForbiddenOnElement(element: Element): boolean {
+      const isNotForbiddenElement = !isAriaHiddenForbiddenOnElement(element);
+      const isPreviousElement = element === previousElement;
+
+      // We came from here
+      if (isPreviousElement) {
+        if (!isNotExcludedElement) {
+          // If any of the ancestors have aria-hidden applied (e.g. by another Modal)
+          // there is a chance that we end up with nothing accessible in the element tree.
+          // So we remove the aria-hidden tag from ancestors so at least the current modal is accessible,
+          // even tho it's probably undesirable when aria-hidden is not coming from another modal.
+          if (show) {
+            ariaHidden(element, !show);
+          }
+          // we restore it if it was originally hidden
+          else {
+            ariaHidden(element, show);
+          }
+        }
```
Can you explain this part of the logic with an example? It's a little hard to follow.

Gr3q replied:
Before this function it only hid siblings. Now it traverses up the html tree from the modal mount location, hiding every sibling and removing aria-hidden from every ancestor of the modal (if needed).

This because models can be mounted anywhere, and we always want to only allow the latest open modal to be accessible.

ZeeshanTamboli commented on packages/mui-material/src/Modal/ModalManager.ts:
```diff
@@ -174,18 +208,32 @@ function handleContainer(containerInfo: Container, props: ManagedModalProps) {
-    }
-  });
+  const html = ownerDocument(container).body.parentElement;
+  let current: Element | null = container;
```
It would be better to change the type of `container` in these methods instead of creating a new variable called `current`.

ZeeshanTamboli replied:
I think this is fine. `parentElement` can be null.

ZeeshanTamboli commented on packages/mui-material/src/Modal/ModalManager.ts:
```diff
@@ -227,8 +281,8 @@ export class ModalManager {
 
-    const hiddenSiblings = getHiddenSiblings(container);
-    ariaHiddenSiblings(container, modal.mount, modal.modalRef, hiddenSiblings, true);
+    const hiddenSiblings = getHiddenElements(container);
```
```suggestion
    const hiddenElements = getHiddenElements(container);
```

ZeeshanTamboli commented on packages/mui-material/src/Modal/useModal.types.ts:
```diff
@@ -44,6 +44,11 @@ export type UseModalParameters = {
    */
   disableEscapeKeyDown?: boolean;
+  /**
+   * The `children` will be under the DOM hierarchy of the parent component.
```
```suggestion
   * If `true`, the `children` will be under the DOM hierarchy of the parent component.
```

ZeeshanTamboli commented on packages/mui-material/src/Portal/Portal.tsx:
```diff
@@ -42,6 +42,7 @@ const Portal = React.forwardRef(function Portal(
 
   useEnhancedEffect(() => {
     if (!disablePortal) {
+      // If you change this the one in useModal.handleOpen->resolvedContainer should match.
```
```suggestion
      // If this is changed, make sure it matches `resolvedContainer` in `useModal.handleOpen`.
```

ZeeshanTamboli reviewed (commented):
@Gr3q Thanks for the pull request.

> In addition because we don't differentiate between aria-hidden tags that got added by modals and added by devs, we need to pierce through (remove) every aria-hidden attribute in every ancestor of the last modal container; this is to prevent no accessible elements on a page when multiple modals are present.

I didn't understand this. Can you please explain more clearly with an example?

> Previous bug still stands when you mess with aria-hidden states in the tree outside your modal, when the modal gets removed your changes won't be preserved (because the modal will restore the state when it was added)

What bug is this? Can you give a bug reproduction in the form of StackBlitz or CodeSandbox?

ZeeshanTamboli reviewed (commented):
@Gr3q, can you address the remaining comments, either by replying or taking action? Also, could you answer my questions [here](https://github.com/mui/material-ui/pull/43318#pullrequestreview-2320235910). The rest of the PR looks good 👍.

We also need to add a test case for the Modal component with the `disablePortal` prop to verify that aria-hidden is handled correctly

Gr3q commented on packages/mui-material/src/Modal/ModalManager.test.ts:
```diff
@@ -386,35 +388,59 @@ describe('ModalManager', () => {
+      modalManager.add(modal2, container2);
+      modalManager.mount(modal2, {});
       // Simulate the main React DOM true.
       expect(container2.children[0]).toBeAriaHidden();
       expect(container2.children[1]).not.toBeAriaHidden();
+      expect(container2.children[2]).toBeAriaHidden();
```
I will change it so I use the elements directly

Gr3q commented on packages/mui-material/src/Modal/ModalManager.test.ts:
```diff
@@ -386,35 +388,59 @@ describe('ModalManager', () => {
+      container2.appendChild(modal2.modalRef);
+      container2.appendChild(modal3.modalRef);
+
+      // ideally mount must be a child of container, and modalRef must be a child of mount.
+      // usually mount is worked out out by Portal:
+      // * if disablePortal is true, mount is modalRef or null (even though types don't allow that).
+      //   it could be the container2 because it is technically the mount (although it doesn't happen in real scenarios).
+      // * if disablePortal is false, mount is the container if set, or document.body.
+
+      // We directly mounted the modalRef to the container2, so
+      // mount is modalRef or the container too (or null, but typing doesn't allow it).
```
I removed the comment it was originally a note to myself on what should I need to handle inside `ModalManager`.

as it says it doesn't happen in real scenarios, but what we usually expect from the modal object is that the DOM order resolves to something like this. `container`->`mountNodeRef`->`modalRef`. From reading the codebase (before my fix) it could happen that `mountNodeRef` resolves to either something unique, `null`, `modalRef` or the `container` itself. So all cases were handled.

After that I added new tests instead of changing this but some changes remained to it, so I reverted them too.

Gr3q commented on packages/mui-material/src/Modal/ModalManager.test.ts:
```diff
@@ -386,35 +388,59 @@ describe('ModalManager', () => {
     });
 
     it('should add aria-hidden to previous modals', () => {
```
removed the changes because they are not needed, see other comment.

Gr3q replied:
Actually it does fail after reverting because `mount` is not a child of `container` and it has to be.

Gr3q replied:
amended the test so it handles all cases [here](https://github.com/mui/material-ui/pull/43318#discussion_r1771126499)

Gr3q commented:
@Gr3q Thanks for the pull request.

 >>In addition because we don't differentiate between aria-hidden tags that got added by modals and added by devs, we need to pierce through (remove) every aria-hidden attribute in every ancestor of the last modal container; this is to prevent no accessible elements on a page when multiple modals are present.

> I didn't understand this. Can you please explain more clearly with an example?

I think I explained the technicality of it [here](https://github.com/mui/material-ui/pull/43318#discussion_r1770523566) but basically someone can put a Modal with `disablePortal` inside an element that is already `aria-hidden` and should be `aria-hidden`, then opens said Modal. In that case we can't afford not to remove that `aria-hidden` tag from the ancestor because we cannot allow to be the whole UI inaccessible, that's the absolute worst case.

If you have multiple modals that function that gets `aria-hidden` elements currently cannot differentiate between if it's actually a dev mistake (and warn) or it was ModalManager that added that prop.

>> Previous bug still stands when you mess with aria-hidden states in the tree outside your modal, when the modal gets removed your changes won't be preserved (because the modal will restore the state when it was added)

> What bug is this? Can you give a bug reproduction in the form of StackBlitz or CodeSandbox?

We get the original elements `aria-hidden` states to restore (were they originally hidden or not) at modal mount time. When we come back and restore them on modal _unmount_, someone might've changed one of those elements while the modal was mounted. In the end we restore them to a state that the dev changed since then. This I think this can only be partially fixed and it would be some effort to implement.

> We also need to add a test case for the Modal component with the disablePortal prop to verify that aria-hidden is handled correctly

I'll add some more tests now.

Gr3q commented:
test added

ZeeshanTamboli commented on packages/mui-material/src/Modal/Modal.test.js:
```diff
@@ -896,4 +896,71 @@ describe('<Modal />', () => {
     expect(handleClose).to.have.property('callCount', 1);
   });
+
+  it('aria-hidden state managed properly with disablePortal', () => {
```
This test and the one below are passing on the `master` branch, meaning they're not testing the correct changes. They should fail on `master` and pass on this branch.

Gr3q replied:
It passed because I didn't test if the modal content was accessible before. On master it isn't accessible (because aria-hidden gets applied to siblings of `body`).

Fixed

ZeeshanTamboli reviewed (commented):
> basically someone can put a Modal with `disablePortal` inside an element that is already `aria-hidden` and should be `aria-hidden`, then opens said Modal. In that case we can't afford not to remove that `aria-hidden` tag from the ancestor because we cannot allow to be the whole UI inaccessible, that's the absolute worst case.

That's the problem of the developer mounting the Modal on the wrong place. We shouldn't handle this and remove `aria-hidden` for them. They should be able to find it out that they are mounting the Modal in the wrong container or remove `aria-hidden` themselves. Could you remove this part of the logic and the relevant test for it (if any).

ZeeshanTamboli reviewed (commented):
@Gr3q, it would’ve been great to use [`aria-modal`](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/#:~:text=Notes%20on%20aria%2Dmodal%20and%20aria%2Dhidden) instead of managing `aria-hidden` ourselves. That would have allowed us to remove  all this logic. Unfortunately, it's not widely supported yet, and also `ariaModal` API [isn't available on Firefox versions below `119`](https://caniuse.com/?search=aria-modal), while [we still support Firefox from version `115`](https://mui.com/material-ui/getting-started/supported-platforms/#browser). Although we don't use `ariaModal` property, can we use `aria-modal` attribute now?

Gr3q commented:
> @Gr3q, it would’ve been great to use [aria-modal](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/#:~:text=Notes%20on%20aria%2Dmodal%20and%20aria%2Dhidden) instead of managing aria-hidden ourselves. That would have allowed us to remove all this logic. Unfortunately, I guess it's not widely supported yet, and also ariaModal API [isn't available on Firefox versions below 119](https://caniuse.com/?search=aria-modal), while [we still support Firefox from version 115](https://mui.com/material-ui/getting-started/supported-platforms/#browser). Although we don't use ariaModal property, do you think we can use aria-modal attribute now?

even though screenreader support is not as good it can still be added. But the docs https://w3c.github.io/aria/#aria-modal implies that there should be only 1 aria-modal at a time, so some logic would need to be written to remove it from all modals excapt the toplevel one.

Plus according the docs we can't escape implementing this same logic because all other elements bust be marked as `inert` (instead of `aria-hidden`). Technically it's not a MUST, but by their spec a modal doesn't enforce focus either.

Quote from the docs

> When a modal element is displayed, authors SHOULD mark all other contents as inert (such as "inert subtrees" in HTML) if the ability to do so exists in the host language.

Gr3q commented:
> That's the problem of the developer mounting the Modal on the wrong place. We shouldn't handle this and remove aria-hidden for them. They should be able to find out that they are mounting the Modal in the wrong container or remove aria-hidden themselves. Could you remove this part of the logic and the relevant test for it (if any).

I was a little too verbose on the dev errors. Taking away dev errors, `ModalManager` adds (and removes) `aria-hidden` tags itself.

Just using that as baseline in this example scenario:
1. User is interacting with a dialog with `disablePortal` enabled
2. The UI has a dialog that opens when it loses connection to the backend without `disablePortal` (so it mounts next to root). The frontend loses connection and the dialog opens while the first dialog was open.

In this case the whole page becomes inaccessible because the dialog the user was using hid the parent element of the lost connection dialog beforehand (if we remove that part of the code).

This is a common enough scenario to handle and even if it isn't I can't guarantee there isn't another edge-case caused by `ModalManager` manipulating `aria-hidden` tags unless you remove `disablePortal` and `container` support from it altogether. So the code must stay.

ZeeshanTamboli reviewed (commented):
@Gr3q, I noticed that in the Menu component (which uses Modal), the scroll isn’t locked when the `disablePortal` prop is used. Here's a video showing the behavior in this PR vs production:

https://github.com/user-attachments/assets/0ddba489-d399-4322-89bd-d1000af8f7f6

`overflow: hidden` style is being applied to the Modal container instead of `document.body` when `disablePortal` is used.

Gr3q commented:
> I noticed that in the Menu component (which uses Modal), the scroll isn’t locked when the disablePortal prop is used. Here's a video showing the behavior in this PR vs production

I see. I'm not sure what is the optimal logic when selecting the `scrollContainer` in `ModalManager.ts#handleContainer`.

This https://github.com/mui/material-ui/pull/33168 changed it so if the modal is in shadow dom it always selects body. Maybe we need to do that in all cases?

ZeeshanTamboli commented:
> Edit: I probably need to ~recursively~ go up the tree to select the correct scrollable ancestor.

Yes, that seems like the right approach. You can use the [`isOverflowing`](https://github.com/mui/material-ui/blob/d72cbe7c36c1097884da75befe06dca2ea174a5a/packages/mui-material/src/Modal/ModalManager.ts#L12) method to check for a vertical scrollbar, and if there isn't, move up using `parentElement`.

Gr3q commented:
I fixed it, the original code was not the best either because it had `isOverflowing` and `containerWindow.getComputedStyle(parent).overflowY === 'scroll'` that more or less do the same thing, but one is used when applied padding and the other one is used to hide the scrollbar.

I could probably refactor this further to do both in 1 loop (aside from the shadow dom fix).

I don't know how to write the tests for this, the expected outcome is that the page and the modal cannot be scrolled.
* when `disabledPortal` is used
* when there are multiple scrollable ancestors
* when there are multiple scrollable ancestors and each of those elements have a fixed element
* when there are multiple scrollable ancestors and each of those elements have a fixed element, and the second (nested) scrollable has `position:relative`

The fixed elements should still look good and be cleaned up properly.   

I know that the current tests are not passing.

Gr3q commented:
> I know that the current tests are not passing.

Those tests are verifying the wrong things, here is a snipped from one:

```typescript
     // This is wrong, this indicates nothing, the page might or might not be scrollable
      expect(document.body.parentElement.style).to.have.property('overflow', '');

      setProps({ open: true });

      // it wasn't scrollable so we didn't need to apply overflow: hidden to it
      expect(document.body.parentElement.style).to.have.property('overflow', 'hidden');

      setProps({ open: false });

      expect(document.body.parentElement.style).to.have.property('overflow', '');
```

ZeeshanTamboli commented on packages/mui-material/src/Modal/ModalManager.ts:
```diff
@@ -120,45 +149,53 @@ function handleContainer(containerInfo: Container, props: ManagedModalProps) {
+      // We need to find all the scrollable ancestors to
+      // prevent scrolling the modal window
+      while (parent != null) {
+        if (containerWindow.getComputedStyle(parent).overflowY === 'scroll') {
```
Why not the HTML check as well from before? (`parent?.nodeName === 'HTML'`)

Gr3q replied:
It was for this

```
// Support html overflow-y: auto for scroll stability between pages
      // https://css-tricks.com/snippets/css/force-vertical-scrollbar/
```

I guess it was to support dev who use this css trick, but the thing is `html` can also be scrollable without this explicit trick (see mui demo). Because I changed the code so it goes up the tree in a loop to find the scrollable items naturally `html` element will be included, so no need for that check especially I'm trying to find all scrollable elements.

Gr3q commented:
I changed the tests for Modal, as I said before they were testing the wrong things, now they are actually checking if their child is mounted or not on open/close.

The remaining tests that are failing in `ModalManager` because of this test (assumption) "should disable the scroll even when not overflowing". I admit I don't understand why is that necessary when the page is not scrollable. Maybe because the Modal content itself can cause an overflow? I can add some code so we always apply `overflow: hidden` to the container as well.

Gr3q commented:
I fixed the behaviour now completely, you can manually add another `overflow:scroll` around the menu button in the docs to test nested scrollable containers. The top bar is also not moving permanently to the left anymore.

The only thing left are the new tests.

ZeeshanTamboli reviewed (commented):
@Gr3q The page remains non-scrollable even when the modal is not opened: https://deploy-preview-43318--material-ui.netlify.app/material-ui/react-modal/

I was thinking it might just be easier to stick with applying `overflow: hidden` on `body` or `html`, even when the modal isn’t in a portal. We’d just keep the same approach as before and override the container to always target `body` or `html`, no matter how deep the modal is nested. Basically, keep it simple and add a comment in the code explaining why we’re doing this override in `handleContainer`. No need of all this logic then. What do you think?

Gr3q commented:
We can do that, the question is  will it work properly if we have a Menu with `disablePortal` inside a Dialog that has scrollable content (which is not a good pattern, but still)?

Gr3q commented:
I reverted back the scrollLock change and amended the test for it. It seems to work fine for both `disabledPortal` and regular modals even with nested scrollbars.

Tests are passing on my end, I can commit `pnpm dedupe` changes but I much prefer you would do that - the tooling for this project gives me trouble on every turn for some reason.

ZeeshanTamboli commented on packages/mui-material/src/Modal/ModalManager.ts:
```diff
@@ -95,23 +114,27 @@ function handleContainer(containerInfo: Container, props: ManagedModalProps) {
+  // We always try to apply the scrollLock to body in case disablePortal was used
+  // on the modal and container is deeply nested. This method seems to block
+  // scrolling on all containers so it still works with disablePortal
+  const body = ownerDocument(containerInfo.container).body;
```
Now that I think of it more I think this break if there are custom overflowing containers provided by the developers through the `container` prop. See https://mui.com/material-ui/api/modal/#modal-prop-container.

Gr3q replied:
I will try it in the docs tomorrow, but when I made the parent container overflowing with `disablePortal` enabled the current implementation still worked.

Gr3q replied:
I checked this by modifying the docs, it seems fine to me. The page is still not scrollable, you can't scroll the passed-in container either. All modals seems to be fine.

I changed the Menu doc to this:

```typescript
import * as React from 'react';
import Button from '@mui/material/Button';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';

export default function BasicMenu() {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const divRef = React.useRef();
  
  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <div ref={divRef} style={{height: "50px", overflowY: "scroll", display: "flex", flexDirection: "column"}}>
        Hello
        Hello
        <span>Hello</span>
        <span>Hello</span>
        <span>Hello</span>
        <span>Hello</span>
      </div>
      <Button
        id="basic-button"
        aria-controls={open ? 'basic-menu' : undefined}
        aria-haspopup="true"
        aria-expanded={open ? 'true' : undefined}
        onClick={handleClick}
      >
        Dashboard
      </Button>
      <Menu
        id="basic-menu"
        anchorEl={anchorEl}
        container={divRef.current}
        open={open}
        onClose={handleClose}
        MenuListProps={{
          'aria-labelledby': 'basic-button',
        }}
      >
        <MenuItem onClick={handleClose}>Profile</MenuItem>
        <MenuItem onClick={handleClose}>My account</MenuItem>
        <MenuItem onClick={handleClose}>Logout</MenuItem>
      </Menu>
    </div>
  );
}
```

ZeeshanTamboli reviewed (commented):
I simplified the logic and tests a bit in https://github.com/mui/material-ui/pull/43318/commits/ba0c158ccfc4b18f4d5ea7c729ac2210bcc563d4, but see my comment below for custom containers. What do you think?

Gr3q commented:
@ZeeshanTamboli any progress on this?

Gr3q commented:
> @Gr3q aria-modal property was added to Dialog in the recent release version 6.1.5 in https://github.com/mui/material-ui/pull/44118.

Yes...and? What do you want me to do about it?

Can I assume you tested screen-reader (including Android, iOS and Windows screen-readers) compatibility, tested multiple open modals at the same time including nested modals? It all works and _usable_, not just compliant?

You can take this PR or leave it, but I definitely don't have time to do any of that for you with `aria-modal`s, that would be just starting from scratch for me with all that testing. Plus I wouldn't be able to remove this patch from our own codebase anyway when I'm done (without good reason), when this implementation passed 3 accessibility audits already.

If you take it, you can build upon it to add/remove `inert` props when they become widely supported (instead of aria-hidden), plus this and https://github.com/mui/material-ui/pull/44118 can exist side-by-side.

You can also leave this PR or take it over completely, I don't mind. In that case I also don't take responsibility for any of the functionality either.

ZeeshanTamboli commented:
> > @Gr3q aria-modal property was added to Dialog in the recent release version 6.1.5 in #44118.
> 
> Yes...and? What do you want me to do about it?
> 
> Can I assume you tested screen-reader (including Android, iOS and Windows screen-readers) compatibility, tested multiple open modals at the same time including nested modals? It all works and _usable_, not just compliant?
> 
> You can take this PR or leave it, but I definitely don't have time to do any of that for you with `aria-modal`s, that would be just starting from scratch for me with all that testing. Plus I wouldn't be able to remove this patch from our own codebase anyway when I'm done (without good reason), when this implementation passed 3 accessibility audits already.
> 
> If you take it, you can build upon it to add/remove `inert` props when they become widely supported (instead of aria-hidden), plus this and #44118 can exist side-by-side.
> 
> You can also leave this PR or take it over completely, I don't mind. In that case I also don't take responsibility for any of the functionality either.

@Gr3q, I just wanted to inform you and see if that could simplify things. No need to start from scratch. 

Could you take a look at this comment: https://github.com/mui/material-ui/pull/43318#discussion_r1816582932?

Gr3q commented:
@Gr3q, I just wanted to inform you and see if that could simplify things. No need to start from scratch.

If you want to keep the previous behaviour - make only the toplevel modal accessible for screenreaders and to be able to trap focus - then I can't.

But I can't say for sure without some real screenreader testing with aria-modals set, like how they handle focus, tabbing, multiple aria-modals, nested aria-modals etc. I could only simplify the code here if they are capable of doing the work for us.

ZeeshanTamboli reviewed (commented):
This page is not scrollable: https://deploy-preview-43318--material-ui.netlify.app/material-ui/react-modal/. Maybe some logic is wrong.

Gr3q commented:
> This page is not scrollable: https://deploy-preview-43318--material-ui.netlify.app/material-ui/react-modal/. Maybe some logic is wrong.

I narrowed it down to the server-side modal on the page. I'll check where it's going wrong.

ZeeshanTamboli commented:
> The server-side modal in the docs is always set to open and `disableScrollLock` is not set on it, so it's technically working as intended. It didn't work as intended before, so technically this is a breaking change?

The issue seems to be that the server-side modal has a custom container (`container={() => rootRef.current!}`). Previously, `handleContainer` considered this custom container, so `overflow: hidden` wasn’t applied to it (because it didn't overflow), and scrolling wasn’t locked. Now, it's always the `body` element, so `overflow: hidden` is applied, locking scrolling. This is the breaking change I was concerned about in [this discussion](https://github.com/mui/material-ui/pull/43318#discussion_r1786644290).

Gr3q commented:
> The issue seems to be that the server-side modal has a custom container (container={() => rootRef.current!}). Previously, handleContainer considered this custom container, so overflow: hidden wasn’t applied to it (because it didn't overflow), and scrolling wasn’t locked. Now, it's always the body element, so overflow: hidden is applied, locking scrolling. This is the breaking change I was concerned about in https://github.com/mui/material-ui/pull/43318#discussion_r1786644290.

@ZeeshanTamboli I'm not sure I have a good answer to that.

My assumption is that `disableScrollLock` is made to prevent scrolling on the page when a modal is open, not on the immediate (parent) container, because that doesn't really have the desired effect (doesn't actually prevent scrolling). Maybe my assumption is wrong.

We made that change because it was not behaving as desired for `disablePortal`. I treat `disablePortal` and  `container` the same, because what you can do with `disablePortal={true}` you can do it directly with `container`. For example in the server-side example if you remove setting `container` it will display exactly the same as with `container`.

If they are meant to be different I'm not sure what is the best course of action, `container` prop has nothing in it's docstring indicating that it will affect scroll locking behaviour in any way.

In short I can't come up with have any use-cases where the behaviour you are describing is useful by myself. If can provide me one or two I can start implementing something that matches those use cases too

ZeeshanTamboli commented:
> @ZeeshanTamboli I'm not sure I have a good answer to that.
> 
> My assumption is that `disableScrollLock` is made to prevent scrolling on the page when a modal is open, not on the immediate (parent) container, because that doesn't really have the desired effect (doesn't actually prevent scrolling). Maybe my assumption is wrong.
>
> We made that change because it was not behaving as desired for `disablePortal`. I treat `disablePortal` and `container` the same, because what you can do with `disablePortal={true}` you can do it directly with `container`. For example in the server-side example if you remove setting `container` it will display exactly the same as with `container`.
> 
> If they are meant to be different I'm not sure what is the best course of action, `container` prop has nothing in it's docstring indicating that it will affect scroll locking behaviour in any way.
> 
> In short I can't come up with have any use-cases where the behaviour you are describing is useful by myself. If can provide me one or two I can start implementing something that matches those use cases too

As per the current production [code](https://github.com/mui/material-ui/blob/master/packages/mui-material/src/Modal/ModalManager.ts#L161), the prevention of scrolling is on the `container`, where the `scrollContainer` in the code is the resolvedContainer (either the one provided by the developer or the default `body`). If you inspect in https://mui.com/material-ui/react-modal/#server-side-modal, you will see `overflow: hidden` is applied on Box container.

You are right that the `disablePortal={true}` is the same as `container`. The `container` prop only has an effect when `disablePortal={false}`. When `disablePortal={true}`, the `container` prop is ignored since no portal is created. But I think it's about the scroll lock style getting applied on the `container`. I think we will have to revert a few commits here till [this comment](https://github.com/mui/material-ui/pull/43318#issuecomment-2382701323).

Gr3q commented:
> > @ZeeshanTamboli I'm not sure I have a good answer to that.
> > My assumption is that `disableScrollLock` is made to prevent scrolling on the page when a modal is open, not on the immediate (parent) container, because that doesn't really have the desired effect (doesn't actually prevent scrolling). Maybe my assumption is wrong.
> > We made that change because it was not behaving as desired for `disablePortal`. I treat `disablePortal` and `container` the same, because what you can do with `disablePortal={true}` you can do it directly with `container`. For example in the server-side example if you remove setting `container` it will display exactly the same as with `container`.
> > If they are meant to be different I'm not sure what is the best course of action, `container` prop has nothing in it's docstring indicating that it will affect scroll locking behaviour in any way.
> > In short I can't come up with have any use-cases where the behaviour you are describing is useful by myself. If can provide me one or two I can start implementing something that matches those use cases too
> 
> As per the current production [code](https://github.com/mui/material-ui/blob/master/packages/mui-material/src/Modal/ModalManager.ts#L161), the prevention of scrolling is on the `container`, where the `scrollContainer` in the code is the resolvedContainer (either the one provided by the developer or the default `body`). If you inspect in https://mui.com/material-ui/react-modal/#server-side-modal, you will see `overflow: hidden` is applied on Box container.
> 
> You are right that the `disablePortal={true}` is the same as `container`. The `container` prop only has an effect when `disablePortal={false}`. When `disablePortal={true}`, the `container` prop is ignored since no portal is created. But I think it's about the scroll lock style getting applied on the `container`. What if a developer has a custom container and no disablePortal? I think we will have to revert a few commits here till [this comment](https://github.com/mui/material-ui/pull/43318#issuecomment-2382701323).

If you want to preserve that behaviour I agree, as long as `disablePortal` will work the same as now in this PR.

You did most of the changes there, would you like me to change it now or would you like to do it?

ZeeshanTamboli commented:
> You did most of the changes there, would you like me to change it now or would you like to do it?

Please look into it. I don't have the bandwidth for it.

Gr3q commented:
> As per the current production [code](https://github.com/mui/material-ui/blob/master/packages/mui-material/src/Modal/ModalManager.ts#L161), the prevention of scrolling is on the container, where the scrollContainer in the code is the resolvedContainer (either the one provided by the developer or the default body). If you inspect in https://mui.com/material-ui/react-modal/#server-side-modal, you will see overflow: hidden is applied on Box container. 
>
> You are right that the disablePortal={true} is the same as container. The container prop only has an effect when disablePortal={false}. When disablePortal={true}, the container prop is ignored since no portal is created. But I think it's about the scroll lock style getting applied on the container. What if a developer has a custom container and no disablePortal?
I think we will have to revert a few commits here till https://github.com/mui/material-ui/pull/43318#issuecomment-2382701323.

I'm changing the code, but I want to summarize what bugs this PR had, then what you want.

Bugs:
1. scrollLock doesn't work for `disablePortal` on the docs page - I found out the correct behaviour is to lock the body element (just like before) because that's the only way scroll locking works. When container is provided locking was applied to the container (and that didn't work, this is the bevaiour on the main branch)
2. Because I changed (fixed) the scroll locking behaviour to always apply to body, the whole modal doc page got locked. This was due to the server-side modal __always__ `open` with no `disableScrollLock`. `disablePortal` applied or not doesn't make a difference here.

So let me clarify what you want:

You want me to change it that scroll lock works peropely when:
* `disablePortal` is set
* `disablePortal` is not set

and it shouldn't work properly (aka scroll lock applies to the container that effectively does nothing to prevent page scrolling):
* `container` is set, regardless if `disablePortal` is set or not. This means even though `container` is not used when `disablePortal` is true and the modal shouldn't be mounted to the container at all, you still want the scroll lock to apply to that container.

I want you to confirm this because this is the only way bugs 1 and 2 can be resolved on the Modal docs page at the same time.

ZeeshanTamboli commented:
@Gr3q It's been a while, but from what I remember you're right—that's what's needed.

Gr3q commented:
@ZeeshanTamboli I fixed the doc page, it does what you want (so no breaking changes).

Added a comment that technically the behaviour around the scrolllock+container is still a bug, hopefully someone will remove it when the next major version comes.

Fixed the existing tests and added new tests for this too.

So now the PR should be ready. :crossed_fingers:

ZeeshanTamboli reviewed (commented):
@Gr3q This looks good to me. I left one comment. 

I'm unsure if this will be considered since the focus is shifting to re-implementing Material UI with the [new Base UI](https://base-ui.com/) in the next major release or the one after. I appreciate the time you've invested in this, and I apologize in advance if it doesn’t move forward. However, it might still be possible to release this now or in the next major version. This will definitely need further review. @DiegoAndai, could you take a look?

Gr3q commented:
> @Gr3q This looks good to me. I left one comment.
> 
> I'm unsure if this will be considered since the focus is shifting to re-implementing Material UI with the [new Base UI](https://base-ui.com/) in the next major release or the one after. I appreciate the time you've invested in this, and I apologize in advance if it doesn’t move forward. However, it might still be possible to release this now or in the next major version. This will definitely need further review. @DiegoAndai, could you take a look?

I think it's still be a while that is released, so it's worth fixing the current version now.

DiegoAndai commented:
Hey @Gr3q and @ZeeshanTamboli. First of all, thanks for working on this 🙌🏼 

> I'm unsure if this will be considered since the focus is shifting to re-implementing Material UI with the [new Base UI](https://base-ui.com/) in the next major release or the one after.

While true, we can still accept and review improvements from the community regarding these topics. So, let's keep working on this. We can land it in v7, which will help many people (given the number of upvotes in the issue).

The first thing I would say is that we should change to using `aria-modal`. Would that be possible? That would remove the need to traverse the DOM by applying/removing `aria-hidden,` which would greatly simplify the implementation.

It will be a breaking change, but we're already on v7 on the `master` branch, so we should make it.

Let me know what you think.

ZeeshanTamboli commented:
> The first thing I would say is that we should change to using `aria-modal`. Would that be possible? That would remove the need to traverse the DOM by applying/removing `aria-hidden,` which would greatly simplify the implementation.

See conversation from https://github.com/mui/material-ui/pull/43318#pullrequestreview-2327122550. That wasn't planned back then but maybe we can consider it now.

DiegoAndai commented on packages/mui-material/src/Modal/useModal.ts:
```diff
@@ -83,7 +85,13 @@ function useModal(parameters: UseModalParameters): UseModalReturnValue {
+     * Resolving this could be simplified (mountNodeRef should take priority when it's set)
+     * but this will also work because the logic matches {@link Portal}.
+     */
+    const resolvedContainer = disablePortal
+      ? ((mountNodeRef.current ?? modalRef.current)?.parentElement ?? getDoc().body)
+      : getContainer(container) || getDoc().body;
```
If providing the `container` prop was required when using `disablePortal`, this could be simplified to 

```suggestion
const resolvedContainer = getContainer(container) || getDoc().body;
```

Right?

Gr3q replied:
But it's not required and it shouldn't be, because if you don't use Portal how could you mount it to a different container?

That is also the reason `disablePortal` takes priority over `container` here.

DiegoAndai reviewed (commented):
Ok, I checked the discussion on this PR and also reached out to the Base UI team.

In conclusion, we are not able to stop using `aria-hidden` as I suggested. So I would say we should move forward with this PR's intention: fix `aria-hidden` handling when `disablePortal` is true.

Here's my initial review.

By the way, thanks for working on this @Gr3q

Gr3q commented on packages/mui-material/src/Modal/ModalManager.ts:
```diff
@@ -56,22 +58,47 @@ function isAriaHiddenForbiddenOnElement(element: Element): boolean {
+      const isPreviousElement = element === previousElement;
+
+      // We came from here
+      if (isPreviousElement) {
```
Yes, that is correct. Can happen if `container` is set to the same thing as the other args.

Gr3q commented on packages/mui-material/src/Modal/ModalManager.ts:
```diff
@@ -95,7 +122,16 @@ function handleContainer(containerInfo: Container, props: ManagedModalProps) {
     value: string;
   }> = [];
-  const container = containerInfo.container;
+  // Implement workaround according to
+  // https://github.com/mui/material-ui/pull/43318#issuecomment-2553509176
+  // Technically applying scrollLock to the container does nothing, but
+  // we preserve the original buggy behavior because fixing it would be
+  // a breaking change.
+  // Original behavior: Apply scroll lock to container if it's set,
+  // otherwise apply to the body element. Because disablePortal
+  // passes in the correct `containerInfo.container`, the easiest way to
+  // make it apply to the correct element is to do this.
+  const container = props.container ? containerInfo.container : document.body;
```
If you do that scroll locking will be broken for `disablePortal` and you will still need to merge them in together.

This PR changed  `containerInfo.container` so it's resolved correctly when `disablePortal` is used, making scroll locking ineffective (because basically it was always applied to body - which works - unless `container` was passed in, where it was also ineffective)
